File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        hash = {}

        for i in range(len(inorder)):
            hash[inorder[i]] = i

        print(hash)

        def fn(inStart, inEnd, postStart, postEnd):
            if (inStart > inEnd or postStart > postEnd):
                return None
            
            root = TreeNode(postorder[postEnd])
            inIdx = hash[postorder[postEnd]]
            rightSize = inEnd - inIdx
            root.left = fn(inStart, inIdx - 1, postStart, postEnd - rightSize - 1)
            root.right = fn(inIdx + 1, inEnd, postEnd - rightSize, postEnd - 1)
            return root
        return fn(0, len(inorder) - 1, 0, len(inorder) - 1)

File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode build (int[] inorder, int[] postorder, int inStart, int inEnd, int postStart, int postEnd, HashMap<Integer, Integer> m) {
        if (inStart > inEnd || postStart > postEnd) {
            return null;
        }
        TreeNode root = new TreeNode(postorder[postEnd]);
        int inIdx = m.get(postorder[postEnd]);
        int rightSize = inEnd - inIdx;  
        root.left = build(inorder, postorder, inStart, inIdx - 1, postStart, postEnd - rightSize - 1, m);
        root.right = build(inorder, postorder, inIdx + 1, inEnd, postEnd - rightSize, postEnd - 1 , m);
        return root;
    }
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        HashMap<Integer, Integer> m = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            m.put(inorder[i], i);
        }
        return build(inorder, postorder, 0, inorder.length - 1, 0, postorder.length - 1, m);
    }
}